[Merged by Bors] - Migrate CI to github actions

lrettig opened the pull request:
## Motivation

I've been struggling to reproduce errors I've seen in travis due to its ultra-wacky run environment (for background see https://gist.github.com/fulldecent/d84cd1abdcc6930865d1b862c4aed917 and https://stackoverflow.com/questions/21053657/how-to-run-travis-ci-locally). I've had more luck with github actions (this tool lets you run them locally: https://github.com/nektos/act). Github actions also makes it very easy to run across an array of platforms.

## Changes
- Mimics travis config, runs all tests using github actions for PRs and push against develop, staging, and trying branches

## To do
- [x] Set up app tests
- [x] Set up system tests
- [x] Set up dockerbuild/push
- [x] Add env vars as repository secrets (done in repo settings, not in this PR)
- [x] Finalize workflow (re-enable unit tests, disable system tests on PR, re-enable system tests on push to staging/trying, etc., see https://github.com/spacemeshos/go-spacemesh/pull/2060#issuecomment-690746582)
- [x] Document workflow and rationale (see https://github.com/spacemeshos/go-spacemesh/pull/2060#issuecomment-690746582)
- [x] Add link from main README
- [x] Disable tests on some paths
- [x] Add slack notification

lrettig committed:
Add basic github actions workflow

lrettig committed:
add make install

lrettig committed:
Add go setup

lrettig committed:
Enable OS matrix, disable unit tests

They're working, but they're slow and they're already running in travis.
For now just try a basic build.

lrettig committed:
pin go version

prevents spurious gofmt/lint errors

lrettig committed:
Don't run install twice, skip lint on windows

lrettig committed:
try again

lrettig committed:
fix quotes

lrettig commented:
Marking this ready for review to indicate that it can be merged. I commented out the full test suite for now since it would duplicate what's already running in Travis. Let's discuss whether we want to transition totally to GH actions, or run this in parallel with Travis. It's nice to have a quick build check across all platforms. As @noamnelke pointed out, this is not currently running system tests that require secrets, but that's also easy to set up in GH actions.

y0sher commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,43 @@
+    #   run: make test-only-app-test
+    - name: fmt, tidy, lint
+      # lint gets confused by windows line endings so skip it on windows
+      if: matrix.os != 'windows-latest'
```
wouldn't it be better to add a windows version or fix the lint to work on both ? it is confusing as until I read this line I was sure it runs perfectly on windows as well..

lrettig replied:
Linting by definition should not be platform-specific though, right? It seems sufficient to run a lint check on a single platform

y0sher reviewed (commented):
I added one note about win support, apart from that I'd love to see at least a issue about converting the UT to github actions as well..

antonlerner commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,43 @@
+    strategy:
+      matrix:
+        # no freebsd support for now =\
+        os: [macos-latest, windows-latest, ubuntu-latest]
```
do we support windows builds?

lrettig replied:
Yes - we have lots of people smeshing on windows

lrettig commented:
We discussed this briefly on Slack, it sounded like everyone is in favor of at least testing GitHub Actions by running CI in parallel along with Travis for a while. Should I re-enable the full test suite here and should we merge this? Do we want to try running the full system test suite here too?

noamnelke commented:
I'm for enabling the full unit test suite here and merging (in parallel with Travis, for now).

I think it's not worth the effort to port the system tests here, because @ilans is working on building full CI capabilities into Spacecraft, so it will be integrated directly with Github. In the meantime we can keep Travis around for that.

lrettig committed:
Fill out all of the tests

Re-enable app tests
Add dockerbuild
Copy over system test config from travis

lrettig committed:
Parameterize go version

lrettig commented:
I went ahead and copied over the config for all of the tests from travis, it was pretty straightforward and anyway per our chat @noamnelke we want to migrate sooner rather than later. App tests are passing for linux but failing for macOS. I'm not sure if we've ever fully, successfully run the test suite on mac (or on Windows) - when I run the p2p tests locally on macOS, I see some failures, and in fact it even crashes my system entirely sometimes. I'll try disabling the other operating systems for now as a test.

lrettig committed:
Just run on ubuntu for now

lrettig committed:
Add more to last commit

lrettig committed:
Fix os matrix syntax

lrettig commented:
I suggest we merge with only linux support, then investigate adding the other operating systems in a new issue/PR. @noamnelke could you look into plugging in bors here?

noamnelke commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,149 @@
+
+# Controls when the action will run. Triggers the workflow on pull requests to a set of branches.
+on:
+  pull_request:
+    branches: [
+      # This is our working branch
+      develop,
+      # This is where pull requests from "bors r+" are built.
+      staging,
+      # This is where pull requests from "bors try" are built. 
+      trying,
+    ]
```
I think all PRs should trigger builds. I actually think that we won't ever have PRs with `staging` or `trying` as their base, but we could legitimately have PRs with other branches as the base.

lrettig replied:
So this isn't required for bors itself? I.e., bors runs the tests some other way? This is how it's configured in travis, I just copied it directly over

noamnelke replied:
hold on... there's two things: `on/pull_request` and `on/push`.

`on/pull_request` you want to run the unit tests for all branches, but `on/push` you only want to run on `staging`, `trying` and `develop`.

`on/push` to `staging` and `trying` you want to run the same things you run `on/pull_request` plus the system tests, while `on/push` to `develop` we only do a push to DockerHub with no tests at all.

I realize this is confusing and I take the blame for the confusion - I wrongly assumed we discussed this. This is not trivial to understand only from looking at the code. I also think my explanation here is not enough and I should probably document it better.

noamnelke commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,149 @@
+  systemtest-latenodes:
+    needs: test
+    if: github.event == 'push' && (github.ref == 'staging' || github.ref == 'trying')
+    runs-on: ${{ matrix.os }}
```
This should never run on more than one platform.

lrettig replied:
You're referring specifically to the `latenodes` test? Out of curiosity, why only this one?

noamnelke replied:
No. All the system tests.

The way system tests work is that there's python code that orchestrates a Kubernetes deployment. We don't need to test to make sure that the python code works on all platforms. The nodes themselves (the actual tested code) always run in Docker containers on linux anyway, so running the same test and changing the platform that the python code runs on is just wasteful - it doesn't achieve anything.

noamnelke commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,149 @@
+    steps:
+    - name: checkout
+      uses: actions/checkout@v2
+    - name: set up go
+      uses: actions/setup-go@v2
```
No need for go... The system tests are written in python. How does this work without installing python, is it there by default?

lrettig replied:
Yes it's there by default, see:

- https://docs.github.com/en/actions/reference/software-installed-on-github-hosted-runners
- https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu1804-README.md

noamnelke replied:
As I wrote one comment prior - the system tests are just python code that orchestrates a Kubernetes deployment. No Go code runs here, so no need to setup Go.

noamnelke reviewed (changes requested):
Note that on Travis we don't run unit tests on push to `develop`. The only thing we run on push to `develop` is push to DockerHub. The reason is that by using bors we ensure that only code that passed all tests is ever merged, so a failure (e.g. due to flakiness) on `develop` is undefined and I prefer not to test in the first place.

lrettig committed:
Factor in review feedback

- Run CI workflow on PR only, all branches
- Only run system tests on a single OS
- No need to set up go for system tests
- Add dockerpush as separate workflow, run only on push to develop

lrettig committed:
Update dockerhub.yml

Small tweaks

lrettig committed:
Add missing env vars from secrets

lrettig committed:
Add docker credentials via secrets

lrettig commented:
@noamnelke I factored in your feedback. I also added the missing env vars as secrets in this repository so they can be used by the workflow. One question:

> Note that on Travis we don't run unit tests on push to develop. The only thing we run on push to develop is push to DockerHub

It seems to me that the current workflow is to run _unit and app tests only_ on PRs, then run _system tests and dockerhub push_ on push to develop - isn't this what we want?

lrettig committed:
Run fmt/tidy/lint first

Running them first will save time/resources since they're much faster

lrettig committed:
Setup env first

lrettig committed:
Attempt to fix secret-related error

See if not echoing from Makefile helps
Temporarily disable unit/app tests to make this faster
Add another missing secret

lrettig committed:
Try removing backslashes

noamnelke commented:
> It seems to me that the current workflow is to run unit and app tests only on PRs, then run system tests and dockerhub push on push to develop - isn't this what we want?

No. I'll explain.

There are two sets of tests we run, one is for convenience (fast feedback) and the other is a "gatekeeper".

The first set of tests is the unit tests (when I say unit tests in all of the comments on this PR I mean all the "cheap" tests which includes `go fmt`, `go mod tidy` and linting) - we run those on every push to a PR.

The second set is a superset of the first, but also includes the expensive system tests. We run the second set (mostly`*`) as a condition for merging a PR into develop. I'll explain how that's implemented, but first some background.

Ideally we'd run the second set on every PR push and block merging until everything passes. However, we can't because it's  prohibitively expensive: each test run like this needs our testing infra exclusively for about an hour. At busy times the queue would explode. We also pay for the infra. Additionally, since anyone on Github can open a PR on our repository, one could theoretically run arbitrary code on our infra if we set it up this way.

So the solution we had before bors was that we ran the system tests manually before merging and had them run automatically after the merge. Developers would sometimes forget to run the tests, or choose not to because they felt their changes shouldn't break anything and then it happened from time to time that the build on `develop` was broken.

Enter bors. Bors manages our merges into `develop`. It only obeys commands from users with push access to the repo, so random devs that open PRs can't run arbitrary code on our infra. But when we tell it we want a PR merged (`bors merge`) it does the following: it makes `staging` be an exact replica of `develop`, it merges the PR into staging and lets the tests run (the second, full set of tests), if they pass - it merges the PR, otherwise it lets us know.

This means that no PR is merged without passing all tests first (incl. system tests), `develop` cannot be broken (in fact we never run any tests *after* a merge, only before) and devs don't need to run system tests manually. It also solves the problem of coordinating who can use the testing infra (so we don't overload it) and has some nice optimizations for optimistically trying to merge several PRs that wait in line together and fall back to an automatic binary search for the offenders if this fails.

So as far as our CI system is concerned, it needs to run the small set of tests on every push to a PR and the large set of tests only on pushes to the `staging` and `trying` branches (not PRs to these branches, but actual pushes to them). The only thing that needs to happen on a push to `develop` is a push to DockerHub (no tests).

I hope this clears things up, sorry for not explaining this before or documenting it better in the code.

---

`*` Why "mostly" as a condition for merging? Because there's also `bors try` which allows us to check proactively before attempting to merge a PR.

lrettig committed:
Remove tty from docker build

Work around error in GH actions: "the input device is not a TTY"
Revert previous change

lrettig commented:
Thanks for the clear explanation @noamnelke, I'll take the opportunity to add this to a new doc as discussed! I'll make sure the tests workflow matches the one you described here before we merge this PR, but for now, I've 1. temporarily disabled the unit tests (since they were passing) and 2. left the system tests running on PRs so that we can test the system tests more quickly and easily. The error they're producing ("Timeout waiting to deployment to be ready") is the same error I'm getting when I try to run them locally. Not yet sure what's causing this, let me know if you have any clue. It could be a kubernetes issue.

noamnelke commented:
There were 8 running deployments on lab1, which would cause the slow deployment setup (and funny enough, it would also cause more stranded deployments, which is a snowball effect).

I've now shut them all down manually. Talk to me when you're online and I'll explain how to do it.

lrettig commented:
It looks like the system test failures we're seeing here are due to the fact that we're trying to run the system tests in a branch with a name other than `develop`, which is hardcoded into the config file in a couple of places, e.g.:

https://github.com/spacemeshos/go-spacemesh/blob/ce9d5da1a17aee64458de361cbe4f85a03d1b49f/tests/sync/genesis/config.yaml#L13

Something is going wrong in the bootstrap process and logs aren't even being generated so I can't tell exactly what's going on. `make dockerbuild-test` generates an image with a different name when it's running in a different branch, so I assume the individual instances/pods aren't finding the image they're looking for. Anyway nothing in this branch touches the system tests or indeed the code at all, so I have no reason to believe they won't pass once this is merged.

lrettig committed:
Update workflow

Run unit tests on all PRs and pushes, run system tests only on push to
staging, trying branches.

lrettig committed:
Clean up docker run interactivity

Add "-t" only in interactive mode (to see test progress)
Compress some redundant args to docker run

lrettig committed:
Only run system tests on push (not PR)

lrettig committed:
Set branch more intelligently in Makefile

lrettig committed:
Allow BRANCH to be manually set/overridden

lrettig commented:
Quick snapshot of where I'm at on this:

- Every time I try to run system tests, locally or in the cloud (GH actions here), it times out waiting for pods to be available. This used to work in the `develop` branch but now it's failing there, too, when I try to run locally. I need to understand this better but it seems like remote pods are trying to run using a docker image `spacemeshos/go-spacemesh-dev:$(BRANCH)` which does not exist in dockerhub for branches other than staging, trying, or develop, which explains part of what's going on, but it does not explain why running these tests in the develop branch fails also.

- There's some fancy Makefile env magic that reads the branch name when running in Travis. We may need to do something similar for GH actions since right now it's seeing only `HEAD` as the branch name. `github.ref` in the env is set to something like `refs/pull/2060/merge` although I think for a `push` (as opposed to a PR) it may get the actual branch name so this will hopefully be pretty straightforward.

lrettig committed:
Run git diff in no pager mode

This is helpful when attached to a tty, e.g., inside `act'

lrettig committed:
Finalize system tests workflow logic

lrettig committed:
Merge branch 'develop' into github-actions

lrettig committed:
No need for github branch var

System tests are now passing when I run locally without this

lrettig committed:
Barebones docs for CI

lrettig committed:
Finish fleshing out basic docs

lrettig committed:
Add link

lrettig committed:
Tweak command

lrettig committed:
Test runners

lrettig committed:
More info on secrets

lrettig commented:
@noamnelke I cleaned up the workflow, made sure everything passes locally (using the `act` tool), and added some basic documentation. Please take another look. From my end this is ready to go.

Does anything else need to be done here to integrate GitHub actions with bors? Also I want it to send notifications to Discord and Slack but I don't have credentials for those (CC @avive).

lrettig committed:
Add link to CI doc to main readme

lrettig committed:
Don't trigger tests if only updating docs

lrettig committed:
Merge branch 'develop' into github-actions

lrettig committed:
Merge branch 'develop' into github-actions

noamnelke commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,99 @@
+         go-version: ${{ env.go-version }}
+     - name: setup env
+       run: make
+     - name: fmt, tidy, lint
+       run: |
+         make genproto
+         make test-tidy
+         make test-fmt
+         make lint
+     - name: unit tests (except app test)
+       run: make test-no-app-test
+     - name: app test
+       run: make test-only-app-test
```
So these run sequentially? This will double the time it takes to run them.

While `fmt, tidy, lint` is quick, the other two sets of tests take 9-10 minutes each -- that's why I've split them and made them run in parallel.

Travis lets us run a set of tests and once they all passed (in parallel) -- start the next set. Is this possible here?

A bigger problem is that we must push to DockerHub prior to running the system tests. The system tests use the DockerHub image to run.

lrettig replied:
Good catches. I made these changes. The only difference compared to the current travis config is that, to keep the workflow simpler, the dockerpush task that runs only on a push to `develop` is in a separate workflow file, with no preconditions (since, as discussed, bors only allows a push to this branch once all tests have passed). I think this is more readable than smushing it into the main CI workflow. Let me know if you disagree.

lrettig committed:
Improve workflow

- run unit tests and app tests in parallel
- add dockerpush before system tests

lrettig committed:
Add back first make step

lrettig committed:
Add slack integration

lrettig commented:
I added Slack integration with Aviv's help, using https://github.com/marketplace/actions/slack-github-actions-slack-integration. There are a bunch of actions in the marketplace for this, for the record here are the others I seriously considered:

- https://github.com/marketplace/actions/slack-notify
- https://github.com/marketplace/actions/slack-notify-build (has the nice feature that it can update a message rather than posting a new one for each status update)

lrettig committed:
Merge branch 'develop' into github-actions

noamnelke reviewed (approved):
Looks great!

Nitpick: The `quicktests` can run in parallel to the `unittests` and `apptests`. This would shave a couple of minutes off the whole process.

lrettig committed:
Run quicktests in parallel to unit, app tests

This will save ~90 secs for each run

lrettig commented:
bors merge

lrettig commented:
bors merge-

lrettig committed:
Bugfix in slack action config

lrettig commented:
bors merge

lrettig commented:
bors try

lrettig commented:
@noamnelke I noticed that bors only reported the travis failures (https://github.com/spacemeshos/go-spacemesh/pull/2060#issuecomment-709577054), even though the GA run failed too. Can we update the bors config somehow to report GA failures? I don't think I have access to this.

lrettig commented:
bors try-

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig commented:
bors merge

lrettig commented:
> @noamnelke I noticed that bors only reported the travis failures ([#2060 (comment)](https://github.com/spacemeshos/go-spacemesh/pull/2060#issuecomment-709577054)), even though the GA run failed too. Can we update the bors config somehow to report GA failures? I don't think I have access to this.

Nevermind, I think I figured this out: #2161
